source/_posts/2025-06-25-piping-in-new-features.md

```diff
@@ -0,0 +1,164 @@
+    |> loadSeveral(...)
+    |> filter(isOnSale(...))
+    |> map(sellWidget(...))
+    |> array_sum(...);
+```
+
+Which... gives us mostly the same thing as the long-discussed scalar methods!  Only pipes are more flexible as you can use any function on the right-side, not just those that have been blessed by the language designers as methods.
+
+At this point, pipe comes very close to being "extension functions", a feature of Kotlin and C# that allows writing functions that look like methods on an object, but are actually just stand-alone functions.  It's spelled a bit differently (`|` instead of `-`), but it's 75% of the way there, for free.
+
+Or take it a step further.  What if some steps in the pipe may return `null`?  We can, with a single function, "lift" the elements of our chain to handle `null` values in the same fashion as null-safe methods.
+
+```php
+function maybe(\Closure $c): \Closure
+{
+	return fn(mixed $arg) => $arg === null ? null : $c($arg);
+}
+
+$profit = [1, 4, 5] 
+    |> maybe(loadSeveral(...))
+    |> maybe(filter(isOnSale(...)))
+    |> maybe(map(sellWidget(...)))
+    |> maybe(array_sum(...));
+```
+
+That's right, we just implemented a Maybe Monad with a pipe and a single-line function.
+
+Now, think about that for streams...
+
+```php
+fopen('pipes.md', 'rb') // No variable, so it will close automatically when GCed.
+    |> decode_rot13(...)
+    |> lines_from_charstream(...)
+    |> map(str_getcsv(...))
+    |> map(Product::create(...))
+    |> map($repo->save(...))
+;
+```
+
+The potential is absolutely huge.  I don't think it's immodest to say that the pipe operator has one of the highest "bangs for the buck" of any feature in recent memory, alongside such niceties as constructor property promotion.  And all thanks to a little syntax sugar.
+
+## Where to from here?
+
+Although pipes are a major milestone, we're not done.  There is active work on not one but two follow up RFCs.  Neither one looks likely to make it into PHP 8.5, but are strong contenders for PHP 8.6 that will fully flesh out what pipes started.
+
+The first is a [function composition operator](https://wiki.php.net/rfc/function-composition).  Where pipe executes immediately, function composition creates a new function by sticking two functions end-to-end.  That would mean the streams example above could be further optimized by combining the `map()` calls:
+
+```php
+fopen('pipes.md', 'rb')
+    |> decode_rot13(...)
+    |> lines_from_charstream(...)
+    |> map(str_getcsv(...) + Product::create(...) + $repo->save(...))
+;
+```
+
+The other is a second attempt at Partial Function Application.  This is a larger lift, but with first-class callables already bringing in much of the necessary plumbing, which simplifies the implementation, and pipes now providing a natural use case, it's worth a second attempt.
+
+I won't promise either one is going to happen.  But they're the next logical step on this journey, so it's worth shooting for.  Stay tuned.
```

Comment:
```suggestion
I won't promise either one is going to happen.  But they're the next logical step on this journey, so it's worth shooting for.  Stay tuned.

> This work was made possible by the support of the PHP Foundation and its sponsors. If you’d like to help push PHP forward, consider [becoming a sponsor](https://thephp.foundation/sponsor/).”
```

Reply:
Well, sort of.  The help from Ilija and Arnaud yes, but I wasn't paid for it. :-)  I do call out the foundation earlier when I mention them.  I think a nod in this direction would make more sense there, in a parenthetical.  That work for you?